pkg/alertmanager/alertmanager.go

```diff
@@ -455,16 +459,54 @@ func (am *Alertmanager) getFullState() (*clusterpb.FullState, error) {
 	return am.state.GetFullState()
 }
 
-// buildIntegrationsMap builds a map of name to the list of integration notifiers off of a
-// list of receiver config.
-func buildIntegrationsMap(nc []config.Receiver, tmpl *template.Template, firewallDialer *util_net.FirewallDialer, logger log.Logger, notifierWrapper func(string, notify.Notifier) notify.Notifier) (map[string][]*notify.Integration, error) {
+// buildIntegrationsMap builds a map of name to the list of integration notifiers off of a list of receiver config.
+func buildIntegrationsMap(nc []*definition.PostableApiReceiver, tmpl *template.Template, firewallDialer *util_net.FirewallDialer, logger log.Logger, notifierWrapper func(string, notify.Notifier) notify.Notifier) (map[string][]*notify.Integration, error) {
+	loggerFactory := newLoggerFactory(logger)
+	whFn := func(n alertingReceivers.Metadata) (alertingReceivers.WebhookSender, error) {
+		return NewSender(logger), nil
+	}
+	emailFn := func(n alertingReceivers.Metadata) (alertingReceivers.EmailSender, error) {
+		return NewSender(logger), nil
+	}
+
 	integrationsMap := make(map[string][]*notify.Integration, len(nc))
 	for _, rcv := range nc {
-		integrations, err := buildReceiverIntegrations(rcv, tmpl, firewallDialer, logger, notifierWrapper)
-		if err != nil {
-			return nil, err
+		// Check if it's a Grafana or an upstream receiver and build the integrations accordingly.
+		if rcv.Type() == definition.GrafanaReceiverType {
+			// The decrypt functions and the context are used to decrypt the configuration.
+			// We don't need to decrypt anything, so we can pass a no-op decrypt func and a context.Background().
+			rCfg, err := alertingNotify.BuildReceiverConfiguration(context.Background(), definition.PostableAPIReceiverToAPIReceiver(rcv), alertingNotify.NoopDecrypt)
+			if err != nil {
+				return nil, err
+			}
+
+			integrations, err := alertingNotify.BuildReceiverIntegrations(
+				rCfg,
+				tmpl,
+				&images.UnavailableProvider{}, // TODO: include images in notifications
+				loggerFactory,
+				whFn,
+				emailFn,
+				1, // orgID is always 1.
+				version.Version,
+			)
+			if err != nil {
+				return nil, err
+			}
+
+			// TODO: use the nfstatus.Integration wrapper for all integrations.
+			upstreamIntegrations := make([]*notify.Integration, 0, len(integrations))
+			for _, integration := range integrations {
+				upstreamIntegrations = append(upstreamIntegrations, integration.Integration())
+			}
```

Comment:
This would be nice pulled out into a function, e.g. `buildGrafanaReceiverIntegrations`